File: tests/ui-tests/src/strategies/user/api.strategy.ts
import type { APIRequestContext } from "@playwright/test";
import type { AuthApi, AuthResult } from "../../api/auth.api";
import { getAdminApiKey } from "../../api/admin.api";
import { UserApi } from "../../api/user.api";
import type { UserWithProfileTestData } from "../../models/user/user.types";
import { AdminUser } from "../../models/user/AdminUser";
import { TestUser } from "../../models/user/TestUser";
import type { UserCreationStrategy } from "./user-creation.strategy";

export function testUserFromAuthResult(
    requestContext: APIRequestContext,
    payload: UserWithProfileTestData,
    response: AuthResult
): TestUser {
    const userApi = new UserApi(requestContext, response.access_token);
    const data: Partial<UserWithProfileTestData> = {
        ...payload,
        ...response.user,
        password: payload.password,
        displayName: response.user.displayName ?? undefined,
    };
    return new TestUser(requestContext, response.access_token, userApi, data);
}

export function adminUserFromAuthResult(
    requestContext: APIRequestContext,
    payload: UserWithProfileTestData,
    response: AuthResult
): AdminUser {
    const userApi = new UserApi(requestContext, response.access_token);
    const data: Partial<UserWithProfileTestData> = {
        ...payload,
        ...response.user,
        password: payload.password,
        displayName: response.user.displayName ?? undefined,
    };
    return new AdminUser(requestContext, response.access_token, userApi, data);
}

export class ApiUserCreationStrategy implements UserCreationStrategy {
    constructor(
        protected readonly requestContext: APIRequestContext,
        protected readonly authApi: AuthApi
    ) {}

    async createUser(user: UserWithProfileTestData): Promise<TestUser> {
        const response = await this.authApi.registerUserWithProfile(user);
        return testUserFromAuthResult(this.requestContext, user, response);
    }
}

/**
 * Creates a user via `POST /auth/admin/register` (bootstrap). Only email, password, and displayName are sent;
 * profile-only fields on {@link UserWithProfileTestData} are ignored by the API. Use {@link ApiUserCreationStrategy}
 * when you need full register-with-profile behavior.
 */
export class AdminApiUserCreationStrategy extends ApiUserCreationStrategy {
    constructor(
        requestContext: APIRequestContext,
        authApi: AuthApi,
        private readonly adminApiKey: string = getAdminApiKey()
    ) {
        super(requestContext, authApi);
    }

    async createUser(user: UserWithProfileTestData): Promise<AdminUser> {
        const response = await this.authApi.createAdmin(
            {
                email: user.email,
                password: user.password,
                displayName: user.displayName,
            },
            this.adminApiKey
        );
        return adminUserFromAuthResult(this.requestContext, user, response);
    }
}

File: tests/backend_tests/src/strategies/user/api_strategy.py
"""
API-backed user creation strategies (mirrors tests/ui-tests/src/strategies/user/api.strategy.ts).

Python uses ``AuthClient`` in place of Playwright ``AuthApi`` + ``APIRequestContext``;
behavior matches ``registerUserWithProfile`` vs ``createAdmin``.
"""

from __future__ import annotations

from typing import Optional

from models.user.registered_user import RegisteredTestUser, registered_test_user_from_auth_result
from models.user.user_types import UserWithProfileTestData
from services.auth_client import AuthClient, get_admin_api_key


class ApiUserCreationStrategy:
    """Registers via POST /auth/register-with-profile (full profile fields supported)."""

    def __init__(self, client: AuthClient) -> None:
        self._client = client

    def create_user(self, user: UserWithProfileTestData) -> RegisteredTestUser:
        auth = self._client.register_with_profile(user)
        return registered_test_user_from_auth_result(user, auth)


class AdminApiUserCreationStrategy(ApiUserCreationStrategy):
    """
    Bootstrap via POST /auth/admin/register — only email, password, displayName are sent;
    profile-only fields on UserWithProfileTestData are ignored by the API.

    Same inheritance shape as TS ``AdminApiUserCreationStrategy extends ApiUserCreationStrategy``.
    """

    def __init__(
        self,
        client: AuthClient,
        admin_api_key: Optional[str] = None,
    ) -> None:
        super().__init__(client)
        self._admin_api_key = admin_api_key

    def create_user(self, user: UserWithProfileTestData) -> RegisteredTestUser:
        key = self._admin_api_key if self._admin_api_key is not None else get_admin_api_key()
        auth = self._client.create_admin(
            email=user.email,
            password=user.password,
            display_name=user.displayName,
            admin_api_key=key,
        )
        return registered_test_user_from_auth_result(user, auth)
